cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/DefaultSystemAppTaskContext.java

```diff
@@ -0,0 +1,182 @@
+import io.cdap.cdap.common.service.RetryStrategy;
+import io.cdap.cdap.common.utils.DirUtils;
+import io.cdap.cdap.internal.app.DefaultPluginConfigurer;
+import io.cdap.cdap.internal.app.runtime.artifact.ArtifactManagerFactory;
+import io.cdap.cdap.internal.app.runtime.artifact.PluginFinder;
+import io.cdap.cdap.internal.app.runtime.plugin.MacroParser;
+import io.cdap.cdap.internal.app.runtime.plugin.PluginInstantiator;
+import io.cdap.cdap.metadata.PreferencesFetcher;
+import io.cdap.cdap.proto.id.NamespaceId;
+import org.apache.twill.discovery.DiscoveryServiceClient;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.Closeable;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Default implementation for {@link io.cdap.cdap.api.service.worker.SystemAppTaskContext}
+ */
+public class DefaultSystemAppTaskContext extends AbstractServiceDiscoverer implements SystemAppTaskContext {
+
+  private static final Logger LOG = LoggerFactory.getLogger(DefaultSystemAppTaskContext.class);
+
+  private final PreferencesFetcher preferencesFetcher;
+  private final CConfiguration cConf;
+  private final ClassLoader artifactClassLoader;
+  private final PluginFinder pluginFinder;
+  private final ArtifactId artifactId;
+  private final DiscoveryServiceClient discoveryServiceClient;
+  private final SecureStore secureStore;
+  private final String artifactNameSpace;
+  private final String serviceName;
+  private final RetryStrategy retryStrategy;
+  private final ArtifactManager artifactManager;
+  private final PluginInstantiator instantiator;
+  private final List<Closeable> closeableList;
+  private io.cdap.cdap.proto.id.ArtifactId protoArtifactId;
+
+  DefaultSystemAppTaskContext(CConfiguration cConf, PreferencesFetcher preferencesFetcher, PluginFinder pluginFinder,
+                              DiscoveryServiceClient discoveryServiceClient, SecureStore secureStore,
+                              String artifactNameSpace, ArtifactId artifactId, ClassLoader artifactClassLoader,
+                              ArtifactManagerFactory artifactManagerFactory, String serviceName) {
+    super(artifactNameSpace);
+    this.cConf = cConf;
+    this.artifactNameSpace = artifactNameSpace;
+    this.serviceName = serviceName;
+    this.retryStrategy = RetryStrategies.fromConfiguration(cConf, Constants.Retry.SERVICE_PREFIX);
+    this.preferencesFetcher = preferencesFetcher;
+    this.artifactClassLoader = artifactClassLoader;
+    this.pluginFinder = pluginFinder;
+    this.artifactId = artifactId;
+    this.discoveryServiceClient = discoveryServiceClient;
+    this.secureStore = secureStore;
+    this.artifactManager = artifactManagerFactory.create(new NamespaceId(artifactNameSpace), retryStrategy);
+    this.closeableList = new ArrayList<>();
+    File pluginsDir = createTempFolder();
```

Comment:
When will this folder getting deleted?

Reply:
Folder is removed when the instantiator is removed.